### ISSUE TITLE
Make at least public types final.

### ISSUE THREAD

**Issue by NightlyNexus** [square/picasso#1781]
My hunch is it'll break all the tests that use Mockito.
Probably have to take it one type at a time.

**Comment by JakeWharton** [comment]
We should make it so that you never _have_ to mock anything, yes, but if you do you mock at layers where the damage to your tests will be minimal.

For example, no one should mock `Picasso` ever. But you can pass in a custom `Picasso` instance with a mock/fake request handler.

Picasso's own tests are a tragedy of black box mocks. I've written some of the worst of the worst style tests up in here. We need to kill them.

**Comment by NightlyNexus** [comment]
yeah, i meant it'll break Picaso's tests. i personally don't care about breaking other people's Mockito tests.

**Comment by NightlyNexus** [comment]
Picasso, Picasso.Builder, RequestCreator, and StatsSnapshot are the public types that are still open.
